crates/bevy_ecs/src/query/access.rs

```diff
@@ -325,6 +354,57 @@ impl<T: SparseSetIndex> FilteredAccess<T> {
     }
 }
 
+// A struct to express something like `Or<(With<A>, With<B>)>`.
+// Filters like `(With<A>, Or<(With<B>, With<C>)>` are expanded into `Or<(With<(A, B)>, With<(B, C)>)>`.
+#[derive(Clone, Eq, PartialEq)]
+struct ExpandedOrWithAccesses {
+    arr: SmallVec<[FixedBitSet; 8]>,
+}
```

Comment:
I'm definitely open to renaming anything related to this struct, these are working names.
I'm also not sure that semantics of `extend`, `union` methods make sense now.